File: 3534.path-existence-queries-in-a-graph-ii.go
#
# @lc app=leetcode id=3534 lang=golang
#
# [3534] Path Existence Queries in a Graph II
#

import "sort"

# @lc code=start
func pathExistenceQueries(n int, nums []int, maxDiff int, queries [][]int) []int {
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	sort.Slice(indices, func(a, b int) bool {
		return nums[indices[a]] < nums[indices[b]]
	})
	pos := make([]int, n)
	for i := range indices {
		pos[indices[i]] = i
	}
	comp := make([]int, n)
	if n > 0 {
		comp[indices[0]] = 0
	}
	cid := 0
	for i := 1; i < n; i++ {
		if nums[indices[i]]-nums[indices[i-1]] > maxDiff {
			cid++
		}
		comp[indices[i]] = cid
	}
	// compute reach
	reach := make([]int, n)
	j := 0
	for i := 0; i < n; i++ {
		for j < n && nums[indices[j]] <= nums[indices[i]]+maxDiff {
			j++
		}
		reach[i] = j - 1
	}
	// doubling table
	const LOG = 18
	par := make([][]int, n)
	for i := range par {
		par[i] = make([]int, LOG)
		par[i][0] = reach[i]
	}
	for d := 1; d < LOG; d++ {
		for i := 0; i < n; i++ {
			p := par[i][d-1]
			par[i][d] = par[p][d-1]
		}
	}
	// queries
	ans := make([]int, len(queries))
	for qi, q := range queries {
		u, v := q[0], q[1]
		if u == v {
			ans[qi] = 0
			continue
		}
		if comp[u] != comp[v] {
			ans[qi] = -1
			continue
		}
		pu := pos[u]
		pv := pos[v]
		if pu > pv {
			pu, pv = pv, pu
		}
		// bin search min k >=1 s.t. reach^k(pu) >= pv
		lo := 1
		hi := n
		for lo < hi {
			mid := lo + (hi-lo)/2
			curr := pu
			k := mid
			for d := 0; d < LOG; d++ {
				if k&1 == 1 {
					curr = par[curr][d]
				}
				k >>= 1
			}
			if curr >= pv {
				hi = mid
			} else {
				lo = mid + 1
			}
		}
		ans[qi] = lo
	}
	return ans
}
# @lc code=end

File: 3534.path-existence-queries-in-a-graph-ii.java
#
# @lc app=leetcode id=3534 lang=java
#
# [3534] Path Existence Queries in a Graph II
#

# @lc code=start
class Solution {
    public int[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        if (n == 0) return new int[0];
        Integer[] indices = new Integer[n];
        for (int i = 0; i < n; i++) {
            indices[i] = i;
        }
        java.util.Arrays.sort(indices, (a, b) -> Integer.compare(nums[a], nums[b]));
        int[] position = new int[n];
        for (int k = 0; k < n; k++) {
            position[indices[k]] = k;
        }
        int[] comp = new int[n];
        comp[0] = 0;
        int cur_comp = 0;
        for (int k = 0; k < n - 1; k++) {
            if (nums[indices[k + 1]] - nums[indices[k]] > maxDiff) {
                cur_comp++;
            }
            comp[k + 1] = cur_comp;
        }
        int[] rght = new int[n];
        int jj = 0;
        for (int i = 0; i < n; i++) {
            while (jj < n && nums[indices[jj]] <= nums[indices[i]] + maxDiff) {
                jj++;
            }
            rght[i] = jj - 1;
        }
        final int LOG = 18;
        int[][] far = new int[LOG][n];
        for (int i = 0; i < n; i++) {
            far[0][i] = rght[i];
        }
        for (int k = 1; k < LOG; k++) {
            for (int i = 0; i < n; i++) {
                int mid = far[k - 1][i];
                far[k][i] = far[k - 1][mid];
            }
        }
        int[] answer = new int[queries.length];
        for (int qi = 0; qi < queries.length; qi++) {
            int u = queries[qi][0];
            int v = queries[qi][1];
            if (u == v) {
                answer[qi] = 0;
                continue;
            }
            int pa = position[u];
            int pb = position[v];
            if (pa > pb) {
                int tmp = pa;
                pa = pb;
                pb = tmp;
            }
            if (comp[pa] != comp[pb]) {
                answer[qi] = -1;
                continue;
            }
            int l = 1;
            int h = pb - pa;
            int res_dist = h;
            while (l <= h) {
                int md = l + (h - l) / 2;
                int cur = pa;
                for (int bk = 0; bk < LOG; bk++) {
                    if ((md & (1 << bk)) != 0) {
                        cur = far[bk][cur];
                    }
                }
                if (cur >= pb) {
                    res_dist = md;
                    h = md - 1;
                } else {
                    l = md + 1;
                }
            }
            answer[qi] = res_dist;
        }
        return answer;
    }
}
# @lc code=end